src/magic_folder/web.py

```diff
@@ -19,6 +21,115 @@
 )
 
 
+def magic_folder_resource(get_magic_folder, get_auth_token, _v1_resource=None):
+    """
+    Create the root resource for the Magic Folder HTTP API.
+
+    :param get_magic_folder: See ``magic_folder_web_service``.
+    :param get_auth_token: See ``magic_folder_web_service``.
+
+    :param IResource _v1_resource: A resource which will take the place of the
+        usual bearer-token-authorized `/v1` resource.  This is intended to
+        make testing easier.
+
+    :return IResource: The resource that is the root of the HTTP API.
+    """
+    if _v1_resource is None:
+        _v1_resource = V1MagicFolderAPI(get_magic_folder)
+
+    root = Resource()
+    root.putChild(
+        b"api",
+        MagicFolderWebApi(get_magic_folder, get_auth_token),
+    )
+    root.putChild(
+        b"v1",
+        BearerTokenAuthorization(
+            _v1_resource,
+            get_auth_token,
+        ),
+    )
+    return root
+
+
+@attr.s
+class BearerTokenAuthorization(Resource, object):
+    """
+    Protect a resource hierarchy with bearer-token based authorization.
+
+    :ivar IResource _resource: The root of a resource hierarchy to which to
+        delegate actual rendering.
+
+    :ivar (IO bytes) _get_auth_token: A function that returns the correct
+        authentication token.
+    """
+    _resource = attr.ib()
+    _get_auth_token = attr.ib()
+
+    def __attrs_post_init__(self):
+        Resource.__init__(self)
+
+    def render(self, request):
+        """
+        Render the wrapped resource if the request carries correct authorization.
+
+        If it does not, render an UNAUTHORIZED response.
+        """
+        if _is_authorized(request, self._get_auth_token):
+            # Authorization checks out, let the protected resource do what it
+            # will.
+            return self._resource.render(request)
+        # Don't let anything through that isn't authorized.
+        return unauthorized(request)
+
+    def getChildWithDefault(self, path, request):
+        """
+        Get the request child from the wrapped resource if the request carries
+        correct authorization.
+
+        If it does not, return an ``Unauthorized`` resource.
+        """
+        if _is_authorized(request, self._get_auth_token):
+            # Authorization checks out, let the protected resource do what it
+            # will.
+            return self._resource.getChildWithDefault(path, request)
+        # Don't let anything through that isn't authorized.
+        return Unauthorized()
+
```

Comment:
Aside: is `getChild` just not-really-used? `Resource`'s implementation seems to simply always return a no-child object. (I guess the ask here is "do we need to implement `getChild` too, or not?)

Reply:
I forgot this and had to go look it up.  `IResource` only includes `getChildWithDefault`.  `getChild` is a convenience helper provided by `Resource` alongside `putChild`.  Real resource traversal calls `getChildWithDefault` which checks to see if there was a matching `putChild` and only calls `getChild` if there wasn't.

So an `IResource` wrapper needs to ignore `getChild` and just deal with `getChildWithDefault`.  `self._resource` might override `getChild` and that's fine as long as `self._resource` is a `Resource` subclass: its `getChildWithDefault` will call its `getChild` if appropriate.